src/main/java/net/sf/mzmine/util/scans/ScanUtils.java

```diff
@@ -795,5 +870,129 @@ public static int getNumberOfSignals(DataPoint[] data, double noiseLevel) {
+    double l = probabilityProductUnnormalized(scanLeft, scanLeft, expectedMassDeviationInPPM, noiseLevel, mzRange);
+    double r = probabilityProductUnnormalized(scanRight, scanRight, expectedMassDeviationInPPM, noiseLevel, mzRange);
+    return d / Math.sqrt(l*r);
+
+  }
+
+  /**
+   * Calculates the probability product without normalization. Usually, this method is only useful if you plan to normalize the spectra (or value) yourself.
+   * @see #probabilityProduct(DataPoint[], DataPoint[], double, double, Range)
+   */
+  public static double probabilityProductUnnormalized(DataPoint[] scanLeft, DataPoint[] scanRight, double expectedMassDeviationInPPM, double noiseLevel, @Nullable  Range<Double> mzRange) {
+    int i, j;
+    double score = 0d;
+    final double allowedDifference = Math.min(1, 1000 * expectedMassDeviationInPPM * 5e-6);
+    final int nl, nr;//=left.length, nr=right.length;
+    if (mzRange==null) {
+      nl = scanLeft.length;
+      nr = scanRight.length;
+      i=0;
+      j=0;
+    } else {
+      nl = findLastPeakWithin(scanLeft,mzRange)+1;
+      nr = findLastPeakWithin(scanRight, mzRange)+1;
+      i = findFirstPeakWithin(scanLeft, mzRange);
+      j = findFirstPeakWithin(scanRight, mzRange);
+      if (i < 0 || j < 0) return 0d;
+    }
+
+    while (i < nl && j < nr) {
+      DataPoint lp = scanLeft[i];
+      if (lp.getIntensity() < noiseLevel) {
+        ++i;
+        continue;
+      }
+      DataPoint rp = scanRight[j];
+      if (rp.getIntensity() < noiseLevel) {
+        ++j;
+        continue;
+      }
+      final double difference = lp.getMZ() - rp.getMZ();
+      if (Math.abs(difference) <= allowedDifference) {
+        final double mzabs = Math.max(200 * expectedMassDeviationInPPM * 1e-6, expectedMassDeviationInPPM * 1e-6 * Math.round(lp.getMZ() + rp.getMZ()) / 2d);
+        final double variance = mzabs * mzabs;
+        double matchScore = probabilityProductScore(lp, rp, variance);
+        score += matchScore;
+        for (int k = i + 1; k < nl; ++k) {
+          DataPoint lp2 = scanLeft[k];
+          final double difference2 = lp2.getMZ() - rp.getMZ();
+          if (Math.abs(difference2) <= allowedDifference) {
+            matchScore = probabilityProductScore(lp2, rp, variance);
+            score += matchScore;
+          } else break;
+        }
+        for (int l = j + 1; l < nr; ++l) {
+          DataPoint rp2 = scanRight[l];
+          final double difference2 = lp.getMZ() - rp2.getMZ();
+          if (Math.abs(difference2) <= allowedDifference) {
+            matchScore = probabilityProductScore(lp, rp2, variance);
+            score += matchScore;
+          } else break;
+        }
+        ++i;
+        ++j;
+      } else if (difference > 0) {
+        ++j;
+
+      } else {
+        ++i;
+      }
+    }
+    return score;
+  }
+  private static double probabilityProductScore(DataPoint lp, DataPoint rp, double variance) {
```

Comment:
Please add some explanation what this method does

Reply:
it's a private method. There is a documentation for probabilityProduct.

Reply:
Even if it is a private method, there will be people looking at the code and trying to understand it after you..